app/src/services/cpk/cpk.ts

```diff
@@ -267,230 +262,31 @@ class CPKService {
-    try {
-      const signer = this.provider.getSigner()
-      const account = await signer.getAddress()
-      const network = await this.provider.getNetwork()
-      const networkId = network.chainId
-
-      const transactions: Transaction[] = []
-
-      const txOptions: TxOptions = {}
-      await this.getGas(txOptions)
-
-      const buyAmount = await this.subRelayFee(amount)
-
-      let collateralAddress
-
-      if (collateral.address === pseudoNativeAssetAddress) {
-        // ultimately WETH will be the collateral if we fund with native ether
-        collateralAddress = getWrapToken(networkId).address
-
-        // we need to send the funding amount in native ether
-        if (!this.isSafeApp) {
-          txOptions.value = buyAmount
-        }
-
-        // Step 0: Wrap ether
-        transactions.push({
-          to: collateralAddress,
-          value: buyAmount.toString(),
-        })
-      } else {
-        collateralAddress = await marketMaker.getCollateralToken()
-      }
-
-      const marketMakerAddress = marketMaker.address
-      const collateralService = new ERC20Service(this.provider, account, collateralAddress)
-      logger.log(`CPK address: ${this.cpk.address}`)
-      const outcomeTokensToBuy = await marketMaker.calcBuyAmount(buyAmount, outcomeIndex)
-      logger.log(`Min outcome tokens to buy: ${outcomeTokensToBuy}`)
-
-      // Check  if the allowance of the CPK to the market maker is enough.
-      const hasCPKEnoughAlowance = await collateralService.hasEnoughAllowance(
-        this.cpk.address,
-        marketMakerAddress,
-        buyAmount,
-      )
-
-      if (!hasCPKEnoughAlowance) {
-        // Step 1:  Approve unlimited amount to be transferred to the market maker)
-        transactions.push({
-          to: collateralAddress,
-          data: ERC20Service.encodeApproveUnlimited(marketMakerAddress),
-        })
-      }
-
-      // Step 2: Transfer the amount of collateral being spent from the user to the CPK
-      // If we are funding with native ether we can skip this step
-      // If we are signed in as a safe we don't need to transfer
-      if (!this.isSafeApp && collateral.address !== pseudoNativeAssetAddress) {
-        // Step 2: Transfer the amount of collateral being spent from the user to the CPK
-        transactions.push({
-          to: collateralAddress,
-          data: ERC20Service.encodeTransferFrom(account, this.cpk.address, buyAmount),
-        })
-      }
-      // Step 3: Buy outcome tokens with the CPK
-      transactions.push({
-        to: marketMakerAddress,
-        data: MarketMakerService.encodeBuy(buyAmount, outcomeIndex, outcomeTokensToBuy),
-      })
+  service = (params: any) => ({ ...params, service: this })
 
+  buyOutcomes = async (params: CPKBuyOutcomesParams): Promise<TransactionReceipt> => {
+    try {
+      const { transactions, txOptions } = await pipe(setup, fee, wrap, approve, transfer, buy)(this.service(params))
```

Comment:
Also Im not sure I 100% understand this service function? Is it taking return parameters of pipe and adding cpk context to it? Would it be possible to do all of that inside base pipe, it would make it much cleaner like that if possible...

Reply:
it takes the input params from `buyOutcomes` and adds `service` to it so we can access the cpk throughout the pipeline, it is definitely ugly! I can't figure out how to inject it in a cleaner way 🤔 has to be another option...

Reply:
think I figured it out, much cleaner now 🙌

Reply:
Awesome fix! This looks so much better!